cloud_AWS/terraform/module/examples/multiple-aws-accounts-multiple-vpc-setup/tf_multi_exec.py

```diff
@@ -0,0 +1,171 @@
+RequestedProfileNames = Optional[List[str]]  # list of profile names matching profiles in ~/.aws/credentials
+
+
+# execute an action for every profile
+def execute_action(action: TerraformAction, profiles: List[AwsProfile]) -> bool:
+    t = Terraform()
+    successful_count = 0
+    for profile in profiles:
+        print(f'Profile: "{profile.name}" ({profile.region})')
+
+        os.environ["AWS_ACCESS_KEY_ID"] = profile.access_key
+        os.environ["AWS_SECRET_ACCESS_KEY"] = profile.secret_key
+        workspace = profile.name  # AWS profiles are mapped to Terraform workspaces
+
+        if validate_workspace_name(workspace) and prepare_workspace(t, workspace) and action(t, profile.region):
+            successful_count += 1
+
+    print(f"Terraform action successfully executed for {successful_count}/{len(profiles)} AWS profile(s).")
+    return successful_count == len(profiles)
+
+
+# create or just switch workspace
+def prepare_workspace(t: Terraform, workspace: str) -> bool:
+    print(f'Preparing TF workspace "{workspace}"')
+    # try switch to workspace
+    return_code, _, _ = t.set_workspace(workspace)
+    if return_code == os.EX_OK:
+        print("Workspace activated")
+        return True
+
+    # probably no such worskpace exists, try to create it
+    return_code, stdout, stderr = t.create_workspace(workspace)
+    if return_code == os.EX_OK:
+        print("Workspace created and activated")
+        return True
+
+    # failure
+    report_tf_output(return_code, stdout, stderr)
+    return False
+
+
+# TerraformAction
+def action_plan(t: Terraform, region: str) -> bool:
+    print("Terraform plan...")
+    code, stdout, stderr = t.plan(var=f"region={region}")
+    report_tf_output(code, stdout, stderr)
+    return code != EX_FAILED
+
+
+# TerraformAction
+def action_apply(t: Terraform, region: str) -> bool:
+    print("Terraform apply...")
+    code, stdout, stderr = t.apply(skip_plan=True, var=f"region={region}")  # skip_plan means auto-approve
+    report_tf_output(code, stdout, stderr)
+    return code != EX_FAILED
+
+
+# TerraformAction
+def action_destroy(t: Terraform, region: str) -> bool:
+    print("Terraform destroy...")
+    code, stdout, stderr = t.apply(destroy=IsFlagged, skip_plan=True, var=f"region={region}")  # auto-approve
+    report_tf_output(code, stdout, stderr)
+    return code != EX_FAILED
+
+
+def validate_workspace_name(workspace: str) -> bool:
+    # workspace name is used as a suffix to certain AWS and Kentik resource names to make them unique
+    # only lowercase alphanumeric characters and hyphens are allowed - S3 bucket name limitation
+
+    is_valid_character = lambda c: c == "-" or (c.isalnum() and c.islower())
+    if not all([is_valid_character(c) for c in workspace]):
+        print(
+            f'Invalid name for workspace: "{workspace}"; must contain only lower case alphanumeric characters and dashes',
```

Comment:
The error message needs to indicate what the user needs to do to make it go away. It may not be obvious that the user actually has to change the name of the AWS profile. I suggest"

```
print(f"'{workspace}' is not acceptable as a Terraform workspace name. Please change the name of corresponding AWS profile to contain only lower case alphanumeric characters and dashes")
```

BTW, would it be possible to allow users to explicitly specify AWS profile to TF workspace mapping on the command-line? Something like `--profiles Prod_Evn_East=prod-env-east,staging`.

Alternatively we can allow to specify the mapping in a config file (e.g. `~/.kentik/aws_tf_workspace_map`).

If we allow user defined mapping, we'd have to verify that workspace names are unique, because the failure mode in case of workspace name collision is too hard to detect.

Reply:
Updating the printout.

Reply:
I think the user doesnt need to know about terraform workspaces, which is an intermediate step, to use the script.
We may avoid making the user providing the names mapping explicitly by transforming the aws profile names using sha1 algorithm; this produces a unique string of lower case alphanumeric characters acceptable as terraform workspace name.
EG d86fec4a9b8b21589692a849a91b78576badb092
I will push example implementation in separate commit

Reply:
I like the idea.